cli/tools/coverage.rs

```diff
@@ -171,392 +172,338 @@ impl CoverageCollector {
   }
 }
 
-enum CoverageReporterKind {
-  Pretty,
-  Lcov,
+struct BranchCoverageItem {
+  line_index: usize,
+  block_number: usize,
+  branch_number: usize,
+  taken: Option<usize>,
+  is_hit: bool,
 }
 
-fn create_reporter(
-  kind: CoverageReporterKind,
-) -> Box<dyn CoverageReporter + Send> {
-  match kind {
-    CoverageReporterKind::Lcov => Box::new(LcovCoverageReporter::new()),
-    CoverageReporterKind::Pretty => Box::new(PrettyCoverageReporter::new()),
-  }
+struct FunctionCoverageItem {
+  name: String,
+  line_index: usize,
+  execution_count: usize,
 }
 
-trait CoverageReporter {
-  fn visit_coverage(
-    &mut self,
-    script_coverage: &ScriptCoverage,
-    script_source: &str,
-    maybe_source_map: Option<Vec<u8>>,
-    maybe_original_source: Option<Arc<String>>,
-  );
-
-  fn done(&mut self);
+struct CoverageReport {
+  url: ModuleSpecifier,
+  file_path: String,
+  named_functions: Vec<FunctionCoverageItem>,
+  branches: Vec<BranchCoverageItem>,
+  found_lines: Vec<(usize, usize)>,
 }
 
-struct LcovCoverageReporter {}
+fn generate_coverage_report(
+  script_coverage: &ScriptCoverage,
+  script_source: &str,
+  maybe_source_map: &Option<Vec<u8>>,
+) -> CoverageReport {
+  let maybe_source_map = maybe_source_map
+    .as_ref()
+    .map(|source_map| SourceMap::from_slice(source_map).unwrap());
+  let text_lines = TextLines::new(script_source);
+
+  let comment_spans = deno_ast::lex(script_source, MediaType::JavaScript)
+    .into_iter()
+    .filter(|item| {
+      matches!(item.inner, deno_ast::TokenOrComment::Comment { .. })
+    })
+    .map(|item| item.span)
+    .collect::<Vec<_>>();
+
+  let url = Url::parse(&script_coverage.url).unwrap();
+  let file_path = url.to_file_path().unwrap();
```

Comment:
Is it guaranteed that these unwraps won't file? Why not make `script_coverage.url` a `ModuleSpecifier`?

Reply:
I just kept the code as-is here for the most part. I don't want to change `script_coverage.url` to a `ModuleSpecifier` for this PR (maybe a followup)... looks like it might be a filename. I will slightly improve this though.